### ISSUE TITLE
Issue with fastInit and kotlin static providers

### ISSUE THREAD

**Issue by fabrantes** [google/dagger#1673]
We're seeing a compilation error when using the fastInit flag. Sample provider:

```
@AssistedModule
@Module(includes = [AssistedInject_SampleActivityProviders::class])
object SampleActivityProviders {

    @Provides
    @Named("SomeName")
    fun provideSomeOptionalString(intent: Intent): String? {
        return intent.getStringExtra(STRING_ID)
    }
    ...
```

The error goes away when we add the `@JvmStatic` annotation to the provider method.

The compilation error is something along the lines of 

```
ERRORS (1)

JAVA COMPILE: DaggerSomeModule.java:xxxx:xx
  Description: non-static method provideSomeOptionalString(Intent) cannot be referenced from a static context
                    local = SampleActivityProviders.provideSomeOptionalString(getIntent());
```

**Comment by chrisjenx** [comment]
Same, just noticed this too only happens on two providers for us:

```
    @Singleton @Provides
    @GooglePlayServices
    fun providesGooglePlayAvailable(@GooglePlayServices status: Int): Boolean = (status == ConnectionResult.SUCCESS)

    @Singleton @Provides
    @GooglePlayServices
    fun providesGooglePlayServicesStatus(app: ISApplication): Int {
        return GoogleApiAvailability.getInstance().isGooglePlayServicesAvailable(app)
    }
```

Looking in our provider and aligning with what @fabrantes put, looks like it only affects primitives.

**Comment by Chang-Eric** [comment]
This sounds like #1648 but I will double check the fix I submitted for fast init.

**Comment by chrisjenx** [comment]
I checked the generated code, looks like primitives try to access the AppModule directly:
```
private int getGooglePlayServicesInteger() {
    Object local = googlePlayServicesInteger;
    if (local instanceof MemoizedSentinel) {
      synchronized (local) {
        local = googlePlayServicesInteger;
        if (local instanceof MemoizedSentinel) {
          local = AppModule.providesGooglePlayServicesStatus(application);
          googlePlayServicesInteger = DoubleCheck.reentrantCheck(googlePlayServicesInteger, local);
        }
      }
    }
    return (int) local;
  }
```

Where as objects will use the provider:
```
private Analytics getAnalytics() {
    Object local = analytics;
    if (local instanceof MemoizedSentinel) {
      synchronized (local) {
        local = analytics;
        if (local instanceof MemoizedSentinel) {
          local = AppModule_ProvidesSegmentFactory.providesSegment(application);
          analytics = DoubleCheck.reentrantCheck(analytics, local);
        }
      }
    }
    return (Analytics) local;
  }
```

**Comment by chrisjenx** [comment]
I actually thought in the case of the "static methods" that it should call the `providesX` methods directly and not generate a `Factory`?

**Comment by Chang-Eric** [comment]
We don't always use the factory as a Factory. It also serves as a place to put methods that just open up the visibility of a method. For example, most `@Provides` methods are not public which means we need to generate a public proxy method on the factory class.

**Comment by bcorso** [comment]
This is a dupe of #1694 and #1694. This should be fixed since [2.25.3](https://github.com/google/dagger/releases/tag/dagger-2.25.3).
